### ISSUE TITLE
Mailbox peripheral reads via dataout are corrupted

### ISSUE THREAD

**Issue by korran** [chipsalliance/caliptra-rtl#45]
These example waveforms use the mailbox via APB, but I also see this behavior from the microcontroller. These waveforms are taken from a simple test case that writes to the mailbox  then reads from the mailbox, executed on verilator.

I am using the RTL from https://github.com/chipsalliance/caliptra-rtl/tree/ahb-mux-fix, which is the same as the main branch but with the fix for #20 cherry-picked.

In this screenshot, you can see the mailbox data being written. The first three words are 0x634ee2f1, 0x59c7ef70, 0x3005c354. Everything looks good and the correct data is being written to the mailbox SRAM.

![Screenshot from 2023-03-30 01-32-00 Annotated](https://user-images.githubusercontent.com/993854/228786159-710b02b0-be13-4bc6-a0e7-933dd0ad8ce5.png)

However, when the test later tries to read using the dataout register, it reads the first three words as 634ee2f1, 634ee2f1, 59c7ef70. You can see that by the time the peripheral reads from address 1 from SRAM, it is too late; it has already returned 634ee2f1 via AHB.

![Screenshot from 2023-03-30 01-32-35 Annotated](https://user-images.githubusercontent.com/993854/228787171-15436031-3dc3-4bfb-a486-bef4cb8bba41.png)

Howard looked at this a bit too, and might have some additional observations.

**Comment by korran** [comment]
Trace is available here:

[trace.45.vcd.zip](https://github.com/chipsalliance/caliptra-rtl/files/11109854/trace.45.vcd.zip)

**Comment by Nitsirks** [comment]
It looks like you are pushing data into the mailbox from SoC, updating the status from uC to pass control back to SoC, and reading back the data from SoC. Is that right?

I believe it's a bug in that path back that we added to allow uC to write a response and SoC to read it. Howard is right that the pre-load of dword 0 is missing there.

We have protocol checkers for the handshake back and forth - but checking the expected data is still a WIP. The data checking we have is only on the initial mailbox contents, so we missed this duplicated dword.

I'm working on a fix now.

**Comment by korran** [comment]
> It looks like you are pushing data into the mailbox from SoC, updating the status from uC to pass control back to SoC, and reading back the data from SoC. Is that right?

It's simpler than that for this reduced test case; APB is reading and writing itself. The uC isn't doing anything with the maibox, it just sets the ready-for-fw bit and then loops forever. The flow is basically this:

```rs
assert!(!apb.soc_mbox.lock().read().lock());
assert!(apb.soc_mbox.lock().read().lock());
assert!(apb.soc_ifc.cptra_flow_status().read().ready_for_fw());
apb.soc_mbox.cmd().write_val(FW_LOAD_CMD_OPCODE);
apb.soc_mbox.dlen().write_val(0x30);
apb.soc_mbox.datain().write_val(0x634ee2f1);
apb.soc_mbox.datain().write_val(0x59c7ef70);
apb.soc_mbox.datain().write_val(0x3005c354);
<snip 9 more writes>
apb.soc_mbox.status().write(|w| w.status(|w| w.data_ready()));
apb.soc_mbox.execute().write(|w| w.execute(true));
assert_eq!(apb.soc_mbox.cmd().read(), FW_LOAD_CMD_OPCODE);
assert_eq!(apb.soc_mbox().dlen().read(), 0x30);
assert!(apb.soc_mbox().status().read().status().data_ready());
assert_eq!(model.soc_mbox().dataout().read(), 0x634ee2f1);

// FAILS: read actually results in 0x634ee2f1
assert_eq!(model.soc_mbox().dataout().read(), 0x59c7ef70); 
```

I also noticed the same corruption problem when running the microcontroller test case here:

https://github.com/chipsalliance/caliptra-sw/blob/1b9b8e7b217f2b6c814c0fa3b47fa3e8fcd29d75/drivers/test-fw/src/bin/mailbox_tests.rs#L59

**Comment by korran** [comment]
hex file for the microcontroller test (WARNING: cannot run without the fix for #20):

[rom-mailbox-failure.hex.zip](https://github.com/chipsalliance/caliptra-rtl/files/11114481/rom-mailbox-failure.hex.zip)


You should see output like this:

```
UART: mailbox::test_try_start_rcv_txn_error...  [ok]
UART: mailbox::test_send_txn_drop...    [ok]
UART: mailbox::test_send_txn_error...   [ok]
UART: mailbox::test_mailbox_loopback... [failed]
UART: Error: panicked at 'assertion failed: (left == right)
UART:   left: [221, 204, 187, 170, 68, 51, 34, 17, 136, 119, 102, 85, 204, 187, 170],
UART:  right: [221, 204, 187, 170, 221, 204, 187, 170, 68, 51, 34, 17, 136, 119, 102]', drivers/test-fw/src/bin/mailbox_tests.rs:102:17
```

**Comment by Nitsirks** [comment]
I was able to reproduce this, working on the fix.

Just to clarify though, are you having the sender/requester set the status before setting execute intentionally?

apb.soc_mbox.status().write(|w| w.status(|w| w.data_ready()));
apb.soc_mbox.execute().write(|w| w.execute(true));

The intent of the status field is to inform the sender/requester of the status of the request. So the receiver is supposed to update the status after the sender/requester sets execute. Receiver would indicate that the requested command is complete if no response is required, or set it to data ready if the uC was supposed to return something.

If you're setting this intentionally to get the FSM to cycle back to EXECUTE_SOC so that the SOC side can read, that's fine. I just wanted to make sure the protocol here was clear.

**Comment by korran** [comment]
> If you're setting this intentionally to get the FSM to cycle back to EXECUTE_SOC so that the SOC side can read, that's fine. I just wanted to make sure the protocol here was clear.

I didn't write this test case, but I suspect that's what @rusty1968 was thinking when he wrote the test case. Unfortunately, @rusty1968's original test used the same word over and over for its mailbox message, so it didn't notice the corruption.

Some context: I first noticed mysterious mailbox corruption when trying to boot the ROM under verilator. Instead of debugging the ROM (which runs very slow), I tweaked @rusty1968's test to use unique words, which got you the waveforms above.

Original APB test (passes under both emulator and verilator):
https://github.com/chipsalliance/caliptra-sw/blob/1b9b8e7b217f2b6c814c0fa3b47fa3e8fcd29d75/hw-model/src/lib.rs#L309

My modified APB test (passes under emulator but fails under verilator):
https://github.com/korran/caliptra-sw/commit/a844b750512c9fe8cec2a73b76f67f2b625c7e41
